Replace monitor implementation with pure Go BPF version

joestringer opened the pull request:
```release-note
Improve performance of cilium monitor
Reduce cilium-agent binary size
```

This PR converts the existing Cilium usage of perf ring buffers from the CGO implementation under `pkg/bpf/perf_linux.go` over to the [pure Go implementation](https://github.com/cilium/ebpf).

Benefits:
* No need to context switch between Go and CGO for each perf message read (=> faster reads)
* Eliminate CGO dependency in `cilium-agent` - reduces binary size from 100MB to 91MB.

[Benchmark on previous implementation](https://github.com/joestringer/cilium/blob/submit/libebpf-monitor%2Bbench/pkg/bpf/perf_linux_test.go):

```
$ sudo go test -check.vv -check.b -tags=privileged_tests -bench=Reader --check.bmem --check.f=Reader
level=info msg="Detected mounted BPF filesystem at /sys/fs/bpf" subsys=bpf
START: perf_linux_test.go:109: BPFPrivilegedTestSuite.BenchmarkReader
PASS: perf_linux_test.go:109: BPFPrivilegedTestSuite.BenchmarkReader
      500000              3852 ns/op             877 B/op         26 allocs/op
    
OK: 1 passed
PASS
ok      github.com/cilium/cilium/pkg/bpf        2.117s
```

[Equivalent benchmark from pure go library](https://github.com/cilium/ebpf/blob/master/perf/reader_test.go):

```
$ sudo go test -bench=Reader ./...
PASS
ok      github.com/cilium/ebpf  0.082s
PASS
ok      github.com/cilium/ebpf/asm      0.003s
PASS
ok      github.com/cilium/ebpf/internal 0.003s
?       github.com/cilium/ebpf/internal/unix    [no test files]
goos: linux
goarch: amd64
pkg: github.com/cilium/ebpf/perf
BenchmarkReader-4         402206              2945 ns/op             472 B/op         14 allocs/op
PASS
ok      github.com/cilium/ebpf/perf     1.858s
```

One minor difference between the existing implementation and this implementation is that the current implementation 

Remaining tasks:
* [x] Add commit to remove CGO from daemon
* [x] Rebase against merged version

Notes:
* This removes the tool for introspecting debug dumps of the ringbuffer. I figure this is in existing v1.6 or earlier branches so we can always revive it if necessary.
* At the moment, this eliminates the timer which calls an error function when no event was received for longer than a predetermined period, code:
  https://github.com/cilium/cilium/blob/525eee50ccc1b942ac8d26956e49f8e61888e6f4/pkg/bpf/perf_linux.go#L495

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9306)
<!-- Reviewable:end -->

joestringer committed:
vendor: Add cilium/ebpf dependency

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
monitor: Rebase against pure Go perf reader

Rework the monitor to run against the new pure golang perf reader
provided by github.com/cilium/ebpf/perf.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Delete perf event reader implementation

Now that we're based against the golang perf event reader, we don't need
this code any more. Delete it!

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/monitor/agent/monitor.go:
```diff
@@ -254,26 +258,13 @@ func (m *Monitor) connectionHandler1_2(parentCtx context.Context, server net.Lis
-	pl := payload.Payload{Data: []byte{}, CPU: c, Lost: el.Lost, Type: payload.RecordLost}
-	m.send(&pl)
+	m.sendLocked(pl)
+	m.Unlock()
```
use defer here

aanm reviewed (commented):
did a 30 sec review ;-)

joestringer committed:
signal: Rebase against pure Go perf reader

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
tools: Remove ring-dump tool

The ring-dump tool was useful in the past to debug issues with dumps
from the perf ring buffer, but now that we're switching the bpf
implementation over to a more generic library I don't see the benefit to
rebasing / maintaining the tool. If necessary we can always pull this
back up from the v1.6 branch.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Disable CGO by default

Now that the perf reader implementation is written in pure Go, we can
disable CGO for the main binary, saving 9% of the main binary size:

Prior to this commit:

  $ ll daemon/cilium-agent
  -rwxrwxr-x 1 joe joe 100M Nov  5 10:05 daemon/cilium-agent

After this commit:

  $ ll daemon/cilium-agent
  -rwxrwxr-x 1 joe joe 91M Nov  5 10:07 daemon/cilium-agent

Signed-off-by: Joe Stringer <joe@cilium.io>
